geoapps/io/validators.py

```diff
@@ -0,0 +1,268 @@
+#  Copyright (c) 2021 Mira Geoscience Ltd.
+#
+#  This file is part of geoapps.
+#
+#  geoapps is distributed under the terms and conditions of the MIT License
+#  (see LICENSE file at the root of this source code package).
+
+from typing import Any, Dict, Tuple
+
+import numpy as np
+
+from .constants import (
+    required_parameters,
+    valid_parameter_keys,
+    valid_parameter_shapes,
+    valid_parameter_types,
+    valid_parameter_values,
+    valid_parameters,
+)
+
+
+class InputValidator:
+    """
+    Validations for inversion parameters.
+
+    Attributes
+    ----------
+    input : dict, optional
+        Input file contents parsed to dict
+
+    Methods
+    -------
+    validate_input(input)
+        Validates input params and contents/type/shape/keys of values.
+    validate(param value)
+        Validates parameter values, types, shapes, and keys.
+
+    """
+
+    def __init__(self, input: Dict[str, Any] = None):
+        self.input = input
+
+    @property
+    def input(self):
+        return self._input
+
+    @input.setter
+    def input(self, val):
+        if val is not None:
+            self.validate_input(val)
+        self._input = val
+
+    def validate_input(self, input: Dict[str, Any]) -> None:
+        """Validates input params and contents/type/shape/keys of values.
+
+        Calls validate method on individual key/value pairs in input, and
+        handles validations requiring knowledge of other parameters.
+
+        Parameters
+        ----------
+        input : dict
```

Comment:
I think @sebhmg would prefer to use the more compact version of `:param name: Decription`. The `type` isn't required technically since already defined above with typing.

Reply:
Same would have to change for all the other functions.

Reply:
true

Reply:
Ok, I'll switch formats.  Good point about the typing.